File: 2025-12-December-GFG-POTD/December 17 - Overlapping Intervals/Solution.py
from typing import List

class Solution:
    def mergeOverlap(self, arr: List[List[int]]) -> List[List[int]]:
        result = []
        
        # Edge case
        if not arr:
            return result

        # Sort intervals by starting time
        arr.sort(key=lambda x: x[0])

        start, end = arr[0]

        for i in range(1, len(arr)):
            if arr[i][0] <= end:
                end = max(end, arr[i][1])
            else:
                result.append([start, end])
                start, end = arr[i]

        # Add the last interval
        result.append([start, end])

        return result

File: 2025-12-December-GFG-POTD/December 17 - Overlapping Intervals/Solution.java
class Solution {
    public ArrayList<int[]> mergeOverlap(int[][] arr) {
        ArrayList<int[]> result = new ArrayList<>();
        
        // Edge case
        if (arr == null || arr.length == 0) {
            return result;
        }

        // Sort intervals by starting time
        Arrays.sort(arr, (a, b) -> Integer.compare(a[0], b[0]));

        // Initialize first interval
        int start = arr[0][0];
        int end = arr[0][1];

        for (int i = 1; i < arr.length; i++) {
            // If current interval overlaps
            if (arr[i][0] <= end) {
                end = Math.max(end, arr[i][1]);
            } else {
                // No overlap, add previous interval
                result.add(new int[]{start, end});
                start = arr[i][0];
                end = arr[i][1];
            }
        }

        // Add the last interval
        result.add(new int[]{start, end});

        return result;
    }
}
